lighthouse-core/audits/seo/tap-targets.js

```diff
@@ -0,0 +1,314 @@
+  getLargestRect,
+} = require('../../lib/rect-helpers');
+const {getTappableRectsFromClientRects} = require('../../lib/tappable-rects');
+
+const FINGER_SIZE_PX = 48;
+// Ratio of the finger area tapping on an unintended element
+// to the finger area tapping on the intended element
+const MAX_ACCEPTABLE_OVERLAP_SCORE_RATIO = 0.25;
+
+
+/**
+ * @param {LH.Artifacts.Rect} cr
+ */
+function clientRectMeetsMinimumSize(cr) {
+  return cr.width >= FINGER_SIZE_PX && cr.height >= FINGER_SIZE_PX;
+}
+
+/**
+ * A target is "too small" if none of it's clientRects are at least the size of a finger.
+ * @param {LH.Artifacts.TapTarget} target
+ */
+function targetIsTooSmall(target) {
+  for (const cr of target.clientRects) {
+    if (clientRectMeetsMinimumSize(cr)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+/**
+ *
+ * @param {LH.Artifacts.TapTarget[]} targets
+ */
+function getTooSmallTargets(targets) {
+  return targets.filter(targetIsTooSmall);
+}
+
+/**
+ *
+ * @param {LH.Artifacts.TapTarget[]} tooSmallTargets
+ * @param {LH.Artifacts.TapTarget[]} allTargets
+ */
+function getAllFailures(tooSmallTargets, allTargets) {
+  /** @type {LH.Audit.TapTargetOverlapDetail[]} */
+  let failures = [];
+
+  tooSmallTargets.forEach(target => {
+    const overlappingTargets = getAllFailuresForTarget(
+      target,
+      allTargets
+    );
+
+    if (overlappingTargets.length > 0) {
+      failures = failures.concat(overlappingTargets);
+    }
+  });
+
+  return failures;
+}
+
+/**
+ *
+ * @param {LH.Artifacts.TapTarget} tapTarget
+ * @param {LH.Artifacts.TapTarget[]} allTapTargets
+ */
+function getAllFailuresForTarget(tapTarget, allTapTargets) {
+  /** @type LH.Audit.TapTargetOverlapDetail[] */
+  const failures = [];
+
+  for (const maybeOverlappingTarget of allTapTargets) {
+    if (maybeOverlappingTarget === tapTarget) {
+      // checking the same target with itself, skip
+      continue;
+    }
+
+    const failure = getFailureForTargetPair(tapTarget, maybeOverlappingTarget);
+    if (failure) {
+      failures.push(failure);
+    }
+  }
+
+  return failures;
+}
+
+/**
+ * @param {LH.Artifacts.TapTarget} tapTarget
+ * @param {LH.Artifacts.TapTarget} maybeOverlappingTarget
+ * @returns {LH.Audit.TapTargetOverlapDetail | null}
+ */
+function getFailureForTargetPair(tapTarget, maybeOverlappingTarget) {
+  // convert client rects to unique tappable areas from a user's perspective
+  const tappableRects = getTappableRectsFromClientRects(tapTarget.clientRects);
+  const isHttpOrHttpsLink = /https?:\/\//.test(tapTarget.href);
+  if (isHttpOrHttpsLink && tapTarget.href === maybeOverlappingTarget.href) {
+    // no overlap because same target action
+    return null;
+  }
+
+  if (allRectsContainedWithinEachOther(tappableRects, maybeOverlappingTarget.clientRects)) {
+    // If one tap target is fully contained within the other that's
+    // probably intentional (e.g. an item with a delete button inside)
+    return null;
```

Comment:
We're generally ok with missing problems and err on the site of false passes. I'd say in this case it's 95% intentional. (And we also avoid failures where the child target has the same action as the parent but we can't tell that it's the same action.)

There's definitely room to improve this, rather than outright ignoring all contained tap targets. If the parent contains a very small child or multiple children that are too close together we'll miss those problems. But I think ignoring those edge cases is an acceptable trade-off against the extra complexity.

I'll add some more comments to the cases where we tolerate potential failures.

Reply:
wfm